[Enhancement] Convert fast-unstake to use StakingInterface, decouplin…

ruseinov opened the pull request:
Resolves https://github.com/paritytech/substrate/issues/12337 or rather what's left of it.
As we decided not to support pool joins at the moment we can probably ignore that requirement?

polkadot companion: https://github.com/paritytech/polkadot/pull/6199

kianenigma commented on primitives/staking/src/lib.rs:
```diff
@@ -135,6 +135,32 @@ pub trait StakingInterface {
 		num_slashing_spans: u32,
 	) -> Result<bool, DispatchError>;
 
+	/// Get StakingLedger by controller account.
```
```suggestion
	/// Get `StakingLedger` by controller account.
```

kianenigma commented on primitives/staking/src/lib.rs:
```diff
@@ -135,6 +135,32 @@ pub trait StakingInterface {
+	/// Get StakingLedger by controller account.
+	fn get_ledger(ctrl: &Self::AccountId) -> StakingLedger<Self>;
+
+	/// The ideal number of staking participants.
```
The doc is very opaque, what is a staking participant? validator? nominator? active validator? validator candidate?

ruseinov replied:
This doc is copied from the staking pallet though, gotta fix both.

kianenigma commented on primitives/staking/src/lib.rs:
```diff
@@ -135,6 +135,32 @@ pub trait StakingInterface {
+	) -> DispatchResult;
+
+	/// Checks whether an account `staker` has been exposed in an era.
+	/// TODO: shall we have this helper here or just expose ErasStakers
```
maybe something to keep in mind: https://github.com/paritytech/ci_cd/issues/626

ruseinov replied:
I try to keep it readable, but yeah, I guess those comments could be longer. Need to find a good tool for intellij

kianenigma commented on primitives/staking/src/lib.rs:
```diff
@@ -135,6 +135,32 @@ pub trait StakingInterface {
+	/// Checks whether an account `staker` has been exposed in an era.
+	/// TODO: shall we have this helper here or just expose ErasStakers
+	/// storage instead for the caller to do as they please?
+	fn is_exposed_in_era(staker: &T::AccountId, era: &EraIndex) -> bool;
```
I think for now this API is fine.

kianenigma replied:
Although -- I take it back. We have `get_ledger`. a similar `get_exposure` would fit nicer next to it.

ruseinov replied:
It's quite a bit different, because we have multiple exposures. 
So do we just return all the exposures that match staker, era? Otherwise it's inefficient.

I think it's best to keep this api tbh, because we are iterating by prefix. Or otherwise `get_exposures` with era.

kianenigma replied:
This is actually fine -- see my comment about `get_ledger`.

ruseinov replied:
That's fine, we don't need eras anywhere else for now, so keeping this signature specific to the use-case sounds good and concise.

ruseinov commented on primitives/staking/src/lib.rs:
```diff
@@ -135,6 +135,32 @@ pub trait StakingInterface {
 	) -> Result<bool, DispatchError>;
 
+	/// Get StakingLedger by controller account.
+	fn get_ledger(ctrl: &Self::AccountId) -> StakingLedger<Self>;
```
@kianenigma We might want to get rid of those, seems redundant now.:
```
	fn active_stake(controller: &Self::AccountId) -> Option<Self::Balance> {
		Self::ledger(controller).map(|l| l.active)
	}

	fn total_stake(controller: &Self::AccountId) -> Option<Self::Balance> {
		Self::ledger(controller).map(|l| l.total)
	}
```

kianenigma replied:
Now I see why the old code was not returning `StakingLedger`. We cannot use `StakingLedger` in `pallet-nomination-pools` or `pallet-fast-unstake` because it is defined in pallet-stakin.

Putting it in a shared crate is also too much of a stretch in my opinion, because these types are arguably an implementation detail.

I think we should keep these, not have `get_ledger` and `get_exposure` and instead create other API function that express a condition, a state or a behavior rather than the explicit type.

ruseinov replied:
Why don't we introduce a special type that will take values out of StakingLedger? Like type `Ledger`. It could even be an alias for now and then change if needed. 

This makes very little sense to keep as is, because:
```
			let ledger =
				pallet_staking::Ledger::<T>::get(&ctrl).ok_or(Error::<T>::NotController)?;
			ensure!(!Queue::<T>::contains_key(&ledger.stash), Error::<T>::AlreadyQueued);
			ensure!(
				Head::<T>::get().map_or(true, |UnstakeRequest { stash, .. }| stash != ledger.stash),
				Error::<T>::AlreadyHead
			);
			// second part of the && is defensive.
			ensure!(
				ledger.active == ledger.total && ledger.unlocking.is_empty(),
				Error::<T>::NotFullyBonded
			);
```

We'd need to call three different methods that operate on exactly the same storage item. Actually no, 4!
At the very least that's not very dry and also a bit of overhead.

kianenigma replied:
my answer to this and many other aspects can be found in https://github.com/paritytech/substrate/compare/ru/feature/fast-unstake-istaking...kiz-ru/feature/fast-unstake-istaking

ruseinov commented on primitives/staking/src/lib.rs:
```diff
@@ -138,7 +138,7 @@ pub trait StakingInterface {
 	fn get_ledger(ctrl: &Self::AccountId) -> StakingLedger<Self>;
 
-	/// The ideal number of staking participants.
+	/// The ideal number of active validators.
```
@kianenigma I think this reflects the situation better than before, judging by usages of this storage value.

kianenigma commented on frame/fast-unstake/src/lib.rs:
```diff
@@ -125,6 +125,9 @@ pub mod pallet {
 		type ControlOrigin: frame_support::traits::EnsureOrigin<Self::RuntimeOrigin>;
 
+		/// The access to staking functionality.
+		type Staking: StakingInterface<Balance = BalanceOf<Self>, AccountId = Self::AccountId>;
```
```suggestion
		type StakingInterface: StakingInterface<Balance = BalanceOf<Self>, AccountId = Self::AccountId>;
```

To be in-line with how `pallet-nomination-pools` is using this.

ruseinov replied:
I actually think that `type Staking` is a better name for it. StakingInterface seems redundant. Done this on purpose to float the idea.

kianenigma replied:
I prefer keeping it consistent with `nomination-pools`. You can chose which. 

That being said, `type Foo: Foo` is a common pattern in Substrate and Rust in general.

ruseinov replied:
Yeah, I guess I'm just not a big fan of the word `Interface` popping up in a type/variable name. It does make sense in a trait, because that's basically `Interface`.

Yep, I have been planning to make the nomination-pools change once the idea has been floated an accepted.

I also like `Staking`, because we will be able to pass `Staking` (pallet) to it and it'll be `Staking = Staking` :)

ruseinov commented:
@kianenigma this needs your attention now. I've done fixes to nomination-pools and fast-unstake, only the tests/benches are failing now, but I'd like to verify the approach first.

ruseinov commented:
bot rebase

kianenigma commented on frame/fast-unstake/Cargo.toml:
```diff
@@ -24,10 +24,6 @@ sp-io = { version = "6.0.0", default-features = false, path = "../../primitives/
 sp-staking = { default-features = false, path = "../../primitives/staking" }
-
-pallet-balances = { default-features = false, path = "../balances" }
-pallet-timestamp = { default-features = false, path = "../timestamp" }
```
oh boy 😮‍💨

ruseinov replied:
Yep, good stuff!

kianenigma commented on frame/nomination-pools/src/mock.rs:
```diff
@@ -123,6 +111,48 @@ impl sp_staking::StakingInterface for StakingMock {
+	}
+
+	fn election_ongoing() -> bool {
+		todo!()
```
```suggestion
		unimplemented!("reason?")
```

ruseinov replied:
Yep, basically we don't need those methods for nom-pools tests, I will make sure this is properly documented via reason.

kianenigma commented on primitives/staking/src/lib.rs:
```diff
@@ -54,67 +56,93 @@ impl<AccountId, Balance> OnStakerSlash<AccountId, Balance> for () {
-/// Trait for communication with the staking pallet.
+/// A struct that reflects stake that an account has in the staking system. Provides a set of
+/// methods to operate on it's properties. Aimed at making `StakingInterface` more concise.
+pub struct Stake<T: StakingInterface + ?Sized> {
```
In the staking pallet, a

`impl From<StakingLedger> for sp_staking::Stake` 

might come handy now.

ruseinov replied:
Can be done, though it's only going to be used in one method in staking pallet, so not sure if this is too helpful or more DRY than it is now

kianenigma reviewed (commented):
Looks overall good, will do a detailed review later. 

Has some leftover low hanging fruit nonetheless.

ruseinov commented on primitives/staking/src/lib.rs:
```diff
@@ -54,67 +56,93 @@ impl<AccountId, Balance> OnStakerSlash<AccountId, Balance> for () {
-	/// This is only guaranteed to reflect the amount locked by the staking system. If there are
-	/// non-staking locks on the bonded pair's balance this may not be accurate.
-	fn total_stake(stash: &Self::AccountId) -> Option<Self::Balance>;
+	/// TODO: Possibly return a result here too for consistency.
```
@kianenigma I think returning a result here is kinda weird, but at the same time if there is some actual error and we just swallow it by returning a false - it might be a bad thing. WDYT?

kianenigma replied:
No strong opinion -- both are fine for me.

kianenigma commented on bin/node/runtime/src/lib.rs:
```diff
@@ -777,7 +779,7 @@ impl pallet_nomination_pools::Config for Runtime {
 	type BalanceToU256 = BalanceToU256;
 	type U256ToBalance = U256ToBalance;
-	type StakingInterface = pallet_staking::Pallet<Self>;
+	type Staking = pallet_staking::Pallet<Self>;
```
```suggestion
	type Staking = Staking; 
```

ruseinov replied:
I was being consistent with the other Staking declarations in this node config. Should we then revisit those too?

kianenigma replied:
No strong opinion, I think historically we use more the type alias rather than `Pallet` directly. 

Here, I suggested it for aesthetics. `X = X` looks better :D 

Personally, I dislike magic keyword that are not defined (`type Staking` is not visible in this scope), so I actually like yours better. 

Beyond the scope of this PR, these are all my thoughts on the matter, do as you prefer.

ruseinov replied:
fixed

kianenigma commented on frame/fast-unstake/src/lib.rs:
```diff
@@ -125,6 +123,21 @@ pub mod pallet {
 
+		/// Just the `Currency::Balance` type; we have this item to allow us to constrain it to
+		/// `From<u64>`.
+		type CurrencyBalance: sp_runtime::traits::AtLeast32BitUnsigned
```
why is this needed?

ruseinov replied:
We actually don't need this anymore afaik, this is just leftover junk I ported over from Staking setup. Going to remove

kianenigma commented on frame/fast-unstake/src/lib.rs:
```diff
@@ -221,28 +234,23 @@ pub mod pallet {
-			let ledger =
-				pallet_staking::Ledger::<T>::get(&ctrl).ok_or(Error::<T>::NotController)?;
-			ensure!(!Queue::<T>::contains_key(&ledger.stash), Error::<T>::AlreadyQueued);
+			let stash = T::Staking::stash_by_ctrl(&ctrl).map_err(|_| Error::<T>::NotController)?;
```
for the sake of posterity, if we were to ignore controllers in this pallet and only allow stash to do operations, things would have been simpler. 

I.e. 

```
The dispatch origin of this call must be signed by the controller account, similar to
		/// `staking::unbond`.
```

is the root of evil. 

Perhaps we should already move to that direction, and either do it.

kianenigma replied:
@rossbulat WDYT?

ruseinov replied:
@kianenigma Probably best to do this move in both staking and fast-unstake to make the code less confusing?
Seeing how fast-unstake is already released with this interface anyway. We could make an issue for this though. 

What is the proposal here, passing the stash account directly?

kianenigma commented on frame/fast-unstake/src/lib.rs:
```diff
@@ -221,28 +234,23 @@ pub mod pallet {
+			ensure!(!Queue::<T>::contains_key(&stash), Error::<T>::AlreadyQueued);
 			ensure!(
-				Head::<T>::get().map_or(true, |UnstakeRequest { stash, .. }| stash != ledger.stash),
+				Head::<T>::get().map_or(true, |UnstakeRequest { stash, .. }| stash != stash),
```
isn't this shadowing the variable stash from the larger scope? 😱 if so, isn't there a compiler warning? why is no test failing? dangerous code. 

either way, please use a different name

ruseinov replied:
It's not shadowing somehow, I thought it's a bit confusing, but looks kinda great :D Yes, will change the name.

kianenigma replied:
would be curious to figure out exactly what is happening here, it seems important and I can see a similar mistake causing damage someday.

ruseinov replied:
It's not a mistake. It works. Because `stash` in this case is a pattern that we match. So it works fine. We can't say `UnstakeRequest { stashAccount .. }` for example, because the field matches exactly.
And this is passing tests. So technically this is not a mistake, but it's confusing. So I will change the outer variable name.

ruseinov replied:
Scratch that, this should actually be shadowing the outer scope. But it does not warn.

ruseinov replied:
Yep, here's the deal:
1. It does shadow, so the equality comparison would always true
2. The reason it does not warn in this case is simply because the shadowing is only happening within `map_or` closure, so there's no `unused_variable` warning, because we use the original variable downstream.

Et voila!

kianenigma commented on frame/nomination-pools/src/lib.rs:
```diff
@@ -622,17 +622,19 @@ impl<T: Config> BondedPool<T> {
 	fn balance_to_point(&self, new_funds: BalanceOf<T>) -> BalanceOf<T> {
-		let bonded_balance =
-			T::StakingInterface::active_stake(&self.bonded_account()).unwrap_or(Zero::zero());
+		let bonded_balance = T::Staking::stake(&self.bonded_account())
```
I guess it is harmless to still have `fn active_balance` and `fn total_balance` on staking interface that are auto-implemented, preventing a little bit of dupe `.map`ing here.

ruseinov replied:
Yeah, I wanted to implement the extra stuff on `Stake` struct, but it'd still require mapping as we work with `Result`. So I guess we'd still be better off with bloating the interface a little.

kianenigma commented on primitives/staking/src/lib.rs:
```diff
@@ -19,8 +19,10 @@
 
 //! A crate which contains primitives that are useful for implementation that uses staking
 //! approaches in general. Definitions related to sessions, slashing, etc go here.
+extern crate core;
```
```suggestion
```

ruseinov commented on frame/fast-unstake/src/lib.rs:
```diff
@@ -401,16 +406,9 @@ pub mod pallet {
+				let remaining = T::Currency::unreserve(&stash, deposit);
 				if !remaining.is_zero() {
 					frame_support::defensive!("`not enough balance to unreserve`");
 					ErasToCheckPerBlock::<T>::put(0);
```
You mean `halt`? Yeah, it makes sense, though since it's already implemented - just another conflict to solve.
I like it though, it's much more DRY and clear.

ruseinov replied:
I guess I'd prefer this to remain in the scope of your batching PR to avoid confusion and duplication across the two. This can be done when this one is merged while rebasing into batching

ruseinov commented:
@kianenigma I have addressed all the comments except for `halt` and the controller stuff.

I think this is outside of the scope of this PR (halt - because it's already implemented in another one, controller - because it's already released in the original version of fast-unstake, so it has little to do with the work we are doing here). 

Let me know if this makes sense.

ruseinov commented on frame/fast-unstake/src/lib.rs:
```diff
@@ -221,28 +222,25 @@ pub mod pallet {
 			);
-			// second part of the && is defensive.
-			ensure!(
-				ledger.active == ledger.total && ledger.unlocking.is_empty(),
```
Nope, we are not exposing the ledger anymore, so the defensive condition has to go.

ruseinov commented:
bot rebase

kianenigma commented on frame/nomination-pools/src/lib.rs:
```diff
@@ -623,7 +623,7 @@ impl<T: Config> BondedPool<T> {
 	fn balance_to_point(&self, new_funds: BalanceOf<T>) -> BalanceOf<T> {
 		let bonded_balance =
-			T::StakingInterface::active_stake(&self.bonded_account()).unwrap_or(Zero::zero());
+			T::Staking::active_balance(&self.bonded_account()).unwrap_or(Zero::zero());
```
```suggestion
			T::Staking::active_stake(&self.bonded_account()).unwrap_or(Zero::zero());
```

kianenigma commented on frame/nomination-pools/src/lib.rs:
```diff
@@ -632,7 +632,7 @@ impl<T: Config> BondedPool<T> {
 	fn points_to_balance(&self, points: BalanceOf<T>) -> BalanceOf<T> {
 		let bonded_balance =
-			T::StakingInterface::active_stake(&self.bonded_account()).unwrap_or(Zero::zero());
+			T::Staking::active_balance(&self.bonded_account()).unwrap_or(Zero::zero());
```
```suggestion
			T::Staking::active_stake(&self.bonded_account()).unwrap_or(Zero::zero());
```

kianenigma commented on frame/nomination-pools/src/lib.rs:
```diff
@@ -738,7 +738,7 @@ impl<T: Config> BondedPool<T> {
 
 		let bonded_balance =
-			T::StakingInterface::active_stake(&self.bonded_account()).unwrap_or(Zero::zero());
+			T::Staking::active_balance(&self.bonded_account()).unwrap_or(Zero::zero());
```
```suggestion
			T::Staking::active_stake(&self.bonded_account()).unwrap_or(Zero::zero());
```

kianenigma commented on frame/nomination-pools/src/lib.rs:
```diff
@@ -2212,7 +2188,7 @@ impl<T: Config> Pallet<T> {
 		debug_assert_eq!(frame_system::Pallet::<T>::consumers(&bonded_account), 0);
 		debug_assert_eq!(
-			T::StakingInterface::total_stake(&bonded_account).unwrap_or_default(),
+			T::Staking::total_balance(&bonded_account).unwrap_or_default(),
```
```suggestion
			T::Staking::total_stake(&bonded_account).unwrap_or_default(),
```

kianenigma commented on frame/nomination-pools/src/tests.rs:
```diff
@@ -79,8 +79,8 @@ fn test_setup_works() {
 		// the bonded_account should be bonded by the depositor's funds.
-		assert_eq!(StakingMock::active_stake(&bonded_account).unwrap(), 10);
-		assert_eq!(StakingMock::total_stake(&bonded_account).unwrap(), 10);
+		assert_eq!(StakingMock::active_balance(&bonded_account).unwrap(), 10);
```
```suggestion
		assert_eq!(StakingMock::active_stake(&bonded_account).unwrap(), 10);
```

kianenigma commented on frame/nomination-pools/src/tests.rs:
```diff
@@ -79,8 +79,8 @@ fn test_setup_works() {
-		assert_eq!(StakingMock::active_stake(&bonded_account).unwrap(), 10);
-		assert_eq!(StakingMock::total_stake(&bonded_account).unwrap(), 10);
+		assert_eq!(StakingMock::active_balance(&bonded_account).unwrap(), 10);
+		assert_eq!(StakingMock::total_balance(&bonded_account).unwrap(), 10);
```
```suggestion
		assert_eq!(StakingMock::total_stake(&bonded_account).unwrap(), 10);
```

kianenigma commented on frame/nomination-pools/src/tests.rs:
```diff
@@ -2368,7 +2368,7 @@ mod unbond {
 			);
 
-			assert_eq!(StakingMock::active_stake(&default_bonded_account()).unwrap(), 0);
+			assert_eq!(StakingMock::active_balance(&default_bonded_account()).unwrap(), 0);
```
```suggestion
			assert_eq!(StakingMock::active_stake(&default_bonded_account()).unwrap(), 0);
```

kianenigma commented on frame/nomination-pools/src/tests.rs:
```diff
@@ -2415,7 +2415,7 @@ mod unbond {
 				);
 
-				assert_eq!(StakingMock::active_stake(&default_bonded_account()).unwrap(), 94);
+				assert_eq!(StakingMock::active_balance(&default_bonded_account()).unwrap(), 94);
```
```suggestion
				assert_eq!(StakingMock::active_stake(&default_bonded_account()).unwrap(), 94);
```

kianenigma commented on frame/nomination-pools/src/tests.rs:
```diff
@@ -2443,7 +2443,7 @@ mod unbond {
 					}
 				);
-				assert_eq!(StakingMock::active_stake(&default_bonded_account()).unwrap(), 2);
+				assert_eq!(StakingMock::active_balance(&default_bonded_account()).unwrap(), 2);
```
```suggestion
				assert_eq!(StakingMock::active_stake(&default_bonded_account()).unwrap(), 2);
```

kianenigma commented on frame/nomination-pools/src/tests.rs:
```diff
@@ -2486,7 +2486,7 @@ mod unbond {
 					}
 				);
-				assert_eq!(StakingMock::active_stake(&default_bonded_account()).unwrap(), 0);
+				assert_eq!(StakingMock::active_balance(&default_bonded_account()).unwrap(), 0);
```
```suggestion
				assert_eq!(StakingMock::active_stake(&default_bonded_account()).unwrap(), 0);
```

kianenigma commented on frame/nomination-pools/src/tests.rs:
```diff
@@ -2614,7 +2614,7 @@ mod unbond {
 					}
 				);
-				assert_eq!(StakingMock::active_stake(&default_bonded_account()).unwrap(), 10);
+				assert_eq!(StakingMock::active_balance(&default_bonded_account()).unwrap(), 10);
```
```suggestion
				assert_eq!(StakingMock::active_stake(&default_bonded_account()).unwrap(), 10);
```

kianenigma commented on frame/nomination-pools/src/tests.rs:
```diff
@@ -2724,7 +2724,7 @@ mod unbond {
 				}
 			);
-			assert_eq!(StakingMock::active_stake(&default_bonded_account()).unwrap(), 0);
+			assert_eq!(StakingMock::active_balance(&default_bonded_account()).unwrap(), 0);
```
```suggestion
			assert_eq!(StakingMock::active_stake(&default_bonded_account()).unwrap(), 0);
```

kianenigma commented on frame/nomination-pools/src/tests.rs:
```diff
@@ -3083,18 +3083,18 @@ mod pool_withdraw_unbonded {
 			// Then there unbonding balance is no longer locked
-			assert_eq!(StakingMock::active_stake(&default_bonded_account()), Some(5));
-			assert_eq!(StakingMock::total_stake(&default_bonded_account()), Some(5));
+			assert_eq!(StakingMock::active_balance(&default_bonded_account()), Ok(5));
+			assert_eq!(StakingMock::total_balance(&default_bonded_account()), Ok(5));
```
```suggestion
			assert_eq!(StakingMock::active_stake(&default_bonded_account()), Ok(5));
			assert_eq!(StakingMock::total_stake(&default_bonded_account()), Ok(5));
```

kianenigma commented on frame/nomination-pools/src/tests.rs:
```diff
@@ -3141,7 +3141,7 @@ mod withdraw_unbonded {
 					StakingMock::set_bonded_balance(
 						default_bonded_account(),
-						StakingMock::active_stake(&default_bonded_account()).unwrap() / 2,
+						StakingMock::active_balance(&default_bonded_account()).unwrap() / 2,
```
```suggestion
						StakingMock::active_stake(&default_bonded_account()).unwrap() / 2,
```

kianenigma commented on frame/nomination-pools/src/tests.rs:
```diff
@@ -3270,7 +3270,7 @@ mod withdraw_unbonded {
 				StakingMock::set_bonded_balance(default_bonded_account(), 300);
 				Balances::make_free_balance_be(&default_bonded_account(), 300);
-				assert_eq!(StakingMock::total_stake(&default_bonded_account()), Some(300));
+				assert_eq!(StakingMock::total_balance(&default_bonded_account()), Ok(300));
```
```suggestion
				assert_eq!(StakingMock::total_stake(&default_bonded_account()), Ok(300));
```

kianenigma commented on frame/nomination-pools/src/tests.rs:
```diff
@@ -4074,12 +4074,15 @@ mod create {
 			assert!(!PoolMembers::<Runtime>::contains_key(11));
-			assert_eq!(StakingMock::active_stake(&next_pool_stash), None);
+			assert_err!(
+				StakingMock::active_balance(&next_pool_stash),
```
```suggestion
				StakingMock::active_stake(&next_pool_stash),
```

kianenigma commented on primitives/staking/src/lib.rs:
```diff
@@ -54,67 +55,95 @@ impl<AccountId, Balance> OnStakerSlash<AccountId, Balance> for () {
-
-	/// Bond some extra amount in the _Stash_'s free balance against the active bonded balance of
-	/// the account. The amount extra actually bonded will never be more than the _Stash_'s free
+	fn active_balance(who: &Self::AccountId) -> Result<Self::Balance, DispatchError> {
```
```suggestion
	fn active_stake(who: &Self::AccountId) -> Result<Self::Balance, DispatchError> {
```

kianenigma commented on primitives/staking/src/lib.rs:
```diff
@@ -54,67 +55,95 @@ impl<AccountId, Balance> OnStakerSlash<AccountId, Balance> for () {
-	/// This is only guaranteed to reflect the amount locked by the staking system. If there are
-	/// non-staking locks on the bonded pair's balance this may not be accurate.
-	fn total_stake(stash: &Self::AccountId) -> Option<Self::Balance>;
+	fn total_balance(who: &Self::AccountId) -> Result<Self::Balance, DispatchError> {
```
```suggestion
	fn total_stake(who: &Self::AccountId) -> Result<Self::Balance, DispatchError> {
```

kianenigma commented:
renamed a few trivial things, LGTM otherwise.

ruseinov commented:
bot merge

ruseinov commented:
bot merge
